tools/journal_proto_processor.hpp

```diff
@@ -0,0 +1,186 @@
+
+  void ProcessRequiredFixed64Field(FieldDescriptor const* descriptor);
+  void ProcessRequiredMessageField(FieldDescriptor const* descriptor);
+  void ProcessRequiredBoolField(FieldDescriptor const* descriptor);
+  void ProcessRequiredDoubleField(FieldDescriptor const* descriptor);
+  void ProcessRequiredInt32Field(FieldDescriptor const* descriptor);
+
+  void ProcessSingleStringField(FieldDescriptor const* descriptor);
+
+  void ProcessOptionalField(FieldDescriptor const* descriptor);
+  void ProcessRepeatedField(FieldDescriptor const* descriptor);
+  void ProcessRequiredField(FieldDescriptor const* descriptor);
+
+  void ProcessField(FieldDescriptor const* descriptor);
+
+  void ProcessInOut(Descriptor const* descriptor,
+                    std::vector<FieldDescriptor const*>* field_descriptors);
+  void ProcessReturn(Descriptor const* descriptor);
+
+  void ProcessMethodExtension(Descriptor const* descriptor);
+
+  // As the recursive methods above traverse the protocol buffer type
+  // declarations, they enter in the following maps (and set) various pieces of
+  // information to help in generating C++ code.  For the simplest use cases
+  // (mostly, the generation of the .hpp file), the values are merely one or
+  // several strings for C++ code snippets.  For more complex use cases (the
+  // generation of the .cpp file) the values are lambdas which transform one or
+  // two C++ code snippets by wrapping them in a more complex structure.
+
+  // The fields that are in-out, i.e. for which fields of the same name exist in
+  // both the In and the Out message.  Note that both fields are present in this
+  // set.  Those fields are transmitted through the interface with an extra
+  // level of indirection.
+  std::set<FieldDescriptor const*> in_out_;
+
+  // For fields that have a (size) option, the name of the size member variable
+  // in the In or Out struct.  Special processing is required when filling those
+  // fields from the struct members.  No data for other fields.
+  std::map<FieldDescriptor const*, std::string> size_member_name_;
+
+  // For all fields, a lambda that takes the name of a local variable containing
+  // data extracted (and deserialized) from the field and returns a list of
+  // expressions to be passed to the interface.  Deals with passing address and
+  // size for fields that have a size member, and with passing by reference for
+  // fields that are in-out or optional.
+  std::map<FieldDescriptor const*,
+           std::function<std::vector<std::string>(
+                             std::string const& identifier)>>
+      field_arguments_fn_;
+
+  // For all fields, a lambda that takes a serialized expression |expr| and a
+  // protocol buffer object |identifier| and returns a statement to assign
+  // |expr| to the proper field of |identifier|.  |identifier| must be a
+  // pointer.  The lambda calls |field_serializer_fn_| to serialize expressions
```

Comment:
Actually, that might get a bit pedantic, since that sort of construct gets used everywhere (for obvious reasons). Maybe comment initially to obviate the confusion?